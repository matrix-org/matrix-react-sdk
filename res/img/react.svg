<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
