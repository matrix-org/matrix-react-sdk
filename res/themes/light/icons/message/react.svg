<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>