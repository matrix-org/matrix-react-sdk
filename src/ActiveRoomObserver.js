/*
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import RoomViewStore from './stores/RoomViewStore';

/**
 * Consumes changes from the RoomViewStore and notifies specific things
 * about when the active room changes. Unlike listening for RoomViewStore
 * changes, you can subscribe to only changes relevant to a particular
 * room.
 *
 * TODO: If we introduce an observer for something else, factor out
 * the adding / removing of listeners & emitting into a common class.
 */
class ActiveRoomObserver {
    constructor() {
        this._listeners = {};

        this._activeRoomId = RoomViewStore.getRoomId();
        // TODO: We could self-destruct when the last listener goes away, or at least
        // stop listening.
        this._roomStoreToken = RoomViewStore.addListener(this._onRoomViewStoreUpdate.bind(this));
    }

    addListener(roomId, listener) {
        if (!this._listeners[roomId]) this._listeners[roomId] = [];
        this._listeners[roomId].push(listener);
    }

    removeListener(roomId, listener) {
        if (this._listeners[roomId]) {
            const i = this._listeners[roomId].indexOf(listener);
            if (i > -1) {
                this._listeners[roomId].splice(i, 1);
            }
        } else {
            console.warn("Unregistering unrecognised listener (roomId=" + roomId + ")");
        }
    }

    _emit(roomId) {
        if (!this._listeners[roomId]) return;

        for (const l of this._listeners[roomId]) {
            l.call();
        }
    }

    _onRoomViewStoreUpdate() {
        // emit for the old room ID
        if (this._activeRoomId) this._emit(this._activeRoomId);

        // update our cache
        this._activeRoomId = RoomViewStore.getRoomId();

        // and emit for the new one
        if (this._activeRoomId) this._emit(this._activeRoomId);
    }
}

if (global.mx_ActiveRoomObserver === undefined) {
    global.mx_ActiveRoomObserver = new ActiveRoomObserver();
}
export default global.mx_ActiveRoomObserver;
