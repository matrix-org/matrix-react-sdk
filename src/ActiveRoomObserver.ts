/*
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import RoomViewStore from './stores/RoomViewStore';

type Listener = (isActive: boolean) => void;

/**
 * Consumes changes from the RoomViewStore and notifies specific things
 * about when the active room changes. Unlike listening for RoomViewStore
 * changes, you can subscribe to only changes relevant to a particular
 * room.
 *
 * TODO: If we introduce an observer for something else, factor out
 * the adding / removing of listeners & emitting into a common class.
 */
export class ActiveRoomObserver {
    private listeners: {[key: string]: Listener[]} = {};
    private _activeRoomId = RoomViewStore.getRoomId();
    private readonly roomStoreToken: string;

    constructor() {
        // TODO: We could self-destruct when the last listener goes away, or at least stop listening.
        this.roomStoreToken = RoomViewStore.addListener(this.onRoomViewStoreUpdate);
    }

    public get activeRoomId(): string {
        return this._activeRoomId;
    }

    public addListener(roomId, listener) {
        if (!this.listeners[roomId]) this.listeners[roomId] = [];
        this.listeners[roomId].push(listener);
    }

    public removeListener(roomId, listener) {
        if (this.listeners[roomId]) {
            const i = this.listeners[roomId].indexOf(listener);
            if (i > -1) {
                this.listeners[roomId].splice(i, 1);
            }
        } else {
            console.warn("Unregistering unrecognised listener (roomId=" + roomId + ")");
        }
    }

    private emit(roomId, isActive: boolean) {
        if (!this.listeners[roomId]) return;

        for (const l of this.listeners[roomId]) {
            l.call(null, isActive);
        }
    }

    private onRoomViewStoreUpdate = () => {
        // emit for the old room ID
        if (this._activeRoomId) this.emit(this._activeRoomId, false);

        // update our cache
        this._activeRoomId = RoomViewStore.getRoomId();

        // and emit for the new one
        if (this._activeRoomId) this.emit(this._activeRoomId, true);
    };
}

if (window.mxActiveRoomObserver === undefined) {
    window.mxActiveRoomObserver = new ActiveRoomObserver();
}
export default window.mxActiveRoomObserver;
