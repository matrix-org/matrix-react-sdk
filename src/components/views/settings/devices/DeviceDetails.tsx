/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';

import { formatDate } from '../../../../DateUtils';
import { _t } from '../../../../languageHandler';
import Heading from '../../typography/Heading';
import { DeviceVerificationStatusCard } from './DeviceVerificationStatusCard';
import { DeviceWithVerification } from './types';

interface Props {
    device: DeviceWithVerification;
}

interface MetadataTable {
    heading?: string;
    values: { label: string, value?: string | React.ReactNode }[];
}

const DeviceDetails: React.FC<Props> = ({ device }) => {
    const metadata: MetadataTable[] = [
        {
            values: [
                { label: _t('Session ID'), value: device.device_id },
                {
                    label: _t('Last activity'),
                    value: device.last_seen_ts && formatDate(new Date(device.last_seen_ts)),
                },
            ],
        },
        {
            heading: _t('Device'),
            values: [
                { label: _t('IP address'), value: device.last_seen_ip },
            ],
        },
    ];
    return <div className='mx_DeviceDetails' data-testid={`device-detail-${device.device_id}`}>
        <section className='mx_DeviceDetails_section'>
            <Heading size='h3'>{ device.display_name ?? device.device_id }</Heading>
            <DeviceVerificationStatusCard device={device} />
        </section>
        <section className='mx_DeviceDetails_section'>
            <p className='mx_DeviceDetails_sectionHeading'>{ _t('Session details') }</p>
            { metadata.map(({ heading, values }, index) => <table
                className='mxDeviceDetails_metadataTable'
                key={index}
            >
                { heading &&
                    <thead>
                        <tr><th>{ heading }</th></tr>
                    </thead>
                }
                <tbody>

                    { values.map(({ label, value }) => <tr key={label}>
                        <td className='mxDeviceDetails_metadataLabel'>{ label }</td>
                        <td className='mxDeviceDetails_metadataValue'>{ value }</td>
                    </tr>) }
                </tbody>
            </table>,
            ) }
        </section>
    </div>;
};

export default DeviceDetails;
