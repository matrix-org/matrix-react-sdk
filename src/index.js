/*
Copyright 2015, 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

var Skinner = require('./Skinner');

var t = require('counterpart');

t.registerTranslations('en', require('../languages/en'));
t.registerTranslations('pt-BR', require('../languages/pt-BR'));

t.setFallbackLocale('en');
t.setLocale(window.navigator.languages[0]);



module.exports.loadSkin = function(skinObject) {
    Skinner.load(skinObject);
};

module.exports.resetSkin = function() {
    Skinner.reset();
};

module.exports.getComponent = function(componentName) {
    return Skinner.getComponent(componentName);
};
