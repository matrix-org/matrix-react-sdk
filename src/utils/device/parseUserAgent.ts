/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import UAParser from "ua-parser-js";

export enum DeviceType {
    Desktop = "Desktop",
    Mobile = "Mobile",
    Web = "Web",
    Unknown = "Unknown",
}
export type ExtendedDeviceInformation = {
    deviceType: DeviceType;
    // eg Google Pixel 6
    deviceModel?: string;
    // eg Android 11
    deviceOperatingSystem?: string;
    // eg Firefox 1.1.0
    client?: string;
};

// Element/1.8.21 (iPhone XS Max; iOS 15.2; Scale/3.00)
const IOS_KEYWORD = "; iOS ";
const BROWSER_KEYWORD = "Mozilla/";

const getDeviceType = (
    userAgent: string,
    device: UAParser.IDevice,
    browser: UAParser.IBrowser,
    operatingSystem: UAParser.IOS,
): DeviceType => {
    if (browser.name === "Electron") {
        return DeviceType.Desktop;
    }
    if (!!browser.name) {
        return DeviceType.Web;
    }
    if (device.type === "mobile" || operatingSystem.name?.includes("Android") || userAgent.indexOf(IOS_KEYWORD) > -1) {
        return DeviceType.Mobile;
    }
    return DeviceType.Unknown;
};

interface CustomValues {
    customDeviceModel?: string;
    customDeviceOS?: string;
}
/**
 * Some mobile model and OS strings are not recognised
 * by the UA parsing library
 * check they exist by hand
 */
const checkForCustomValues = (userAgent: string): CustomValues => {
    if (userAgent.includes(BROWSER_KEYWORD)) {
        return {};
    }

    const mightHaveDevice = userAgent.includes("(");
    if (!mightHaveDevice) {
        return {};
    }
    const deviceInfoSegments = userAgent.substring(userAgent.indexOf("(") + 1).split("; ");
    const customDeviceModel = deviceInfoSegments[0] || undefined;
    const customDeviceOS = deviceInfoSegments[1] || undefined;
    return { customDeviceModel, customDeviceOS };
};

const concatenateNameAndVersion = (name?: string, version?: string): string | undefined =>
    name && [name, version].filter(Boolean).join(" ");

export const parseUserAgent = (userAgent?: string): ExtendedDeviceInformation => {
    if (!userAgent) {
        return {
            deviceType: DeviceType.Unknown,
        };
    }

    const parser = new UAParser(userAgent);

    const browser = parser.getBrowser();
    const device = parser.getDevice();
    const operatingSystem = parser.getOS();

    const deviceType = getDeviceType(userAgent, device, browser, operatingSystem);

    // OSX versions are frozen at 10.15.17 in UA strings https://chromestatus.com/feature/5452592194781184
    // ignore OS version in browser based sessions
    const shouldIgnoreOSVersion = deviceType === DeviceType.Web || deviceType === DeviceType.Desktop;
    const deviceOperatingSystem = concatenateNameAndVersion(
        operatingSystem.name,
        shouldIgnoreOSVersion ? undefined : operatingSystem.version,
    );
    const deviceModel = concatenateNameAndVersion(device.vendor, device.model);
    const client = concatenateNameAndVersion(browser.name, browser.version);

    // only try to parse custom model and OS when device type is known
    const { customDeviceModel, customDeviceOS } =
        deviceType !== DeviceType.Unknown ? checkForCustomValues(userAgent) : ({} as CustomValues);

    return {
        deviceType,
        deviceModel: deviceModel || customDeviceModel,
        deviceOperatingSystem: deviceOperatingSystem || customDeviceOS,
        client,
    };
};
