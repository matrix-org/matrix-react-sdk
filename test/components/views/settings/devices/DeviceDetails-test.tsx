/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { ComponentProps } from "react";
import { fireEvent, render } from "@testing-library/react";
import { PUSHER_ENABLED } from "matrix-js-sdk/src/matrix";

import DeviceDetails from "../../../../../src/components/views/settings/devices/DeviceDetails";
import { mkPusher } from "../../../../test-utils/test-utils";
import { DeviceType } from "../../../../../src/utils/device/parseUserAgent";

describe("<DeviceDetails />", () => {
    const baseDevice = {
        device_id: "my-device",
        isVerified: false,
        deviceType: DeviceType.Unknown,
    };
    const defaultProps: ComponentProps<typeof DeviceDetails> = {
        device: baseDevice,
        isSigningOut: false,
        onSignOutDevice: jest.fn(),
        saveDeviceName: jest.fn(),
        setPushNotifications: jest.fn(),
        supportsMSC3881: true,
    };

    const getComponent = (props = {}) => <DeviceDetails {...defaultProps} {...props} />;

    // 14.03.2022 16:15
    const now = 1647270879403;
    jest.useFakeTimers();

    beforeEach(() => {
        jest.setSystemTime(now);
    });

    it("renders device without metadata", () => {
        const { container } = render(getComponent());
        expect(container).toMatchSnapshot();
    });

    it("renders device with metadata", () => {
        const device = {
            ...baseDevice,
            display_name: "My Device",
            last_seen_ip: "123.456.789",
            last_seen_ts: now - 60000000,
            appName: "Element Web",
            client: "Firefox 100",
            deviceModel: "Iphone X",
            deviceOperatingSystem: "Windows 95",
        };
        const { container } = render(getComponent({ device }));
        expect(container).toMatchSnapshot();
    });

    it("renders a verified device", () => {
        const device = {
            ...baseDevice,
            isVerified: true,
        };
        const { container } = render(getComponent({ device }));
        expect(container).toMatchSnapshot();
    });

    it("disables sign out button while sign out is pending", () => {
        const device = {
            ...baseDevice,
        };
        const { getByTestId } = render(getComponent({ device, isSigningOut: true }));
        expect(getByTestId("device-detail-sign-out-cta").getAttribute("aria-disabled")).toEqual("true");
    });

    it("renders the push notification section when a pusher exists", () => {
        const device = {
            ...baseDevice,
        };
        const pusher = mkPusher({
            device_id: device.device_id,
        });

        const { getByTestId } = render(
            getComponent({
                device,
                pusher,
                isSigningOut: true,
            }),
        );

        expect(getByTestId("device-detail-push-notification")).toBeTruthy();
    });

    it("hides the push notification section when no pusher", () => {
        const device = {
            ...baseDevice,
        };

        const { getByTestId } = render(
            getComponent({
                device,
                pusher: null,
                isSigningOut: true,
            }),
        );

        expect(() => getByTestId("device-detail-push-notification")).toThrow();
    });

    it("disables the checkbox when there is no server support", () => {
        const device = {
            ...baseDevice,
        };
        const pusher = mkPusher({
            device_id: device.device_id,
            [PUSHER_ENABLED.name]: false,
        });

        const { getByTestId } = render(
            getComponent({
                device,
                pusher,
                isSigningOut: true,
                supportsMSC3881: false,
            }),
        );

        const checkbox = getByTestId("device-detail-push-notification-checkbox");

        expect(checkbox.getAttribute("aria-disabled")).toEqual("true");
        expect(checkbox.getAttribute("aria-checked")).toEqual("false");
    });

    it("changes the pusher status when clicked", () => {
        const device = {
            ...baseDevice,
        };

        const enabled = false;

        const pusher = mkPusher({
            device_id: device.device_id,
            [PUSHER_ENABLED.name]: enabled,
        });

        const { getByTestId } = render(
            getComponent({
                device,
                pusher,
                isSigningOut: true,
            }),
        );

        const checkbox = getByTestId("device-detail-push-notification-checkbox");

        fireEvent.click(checkbox);

        expect(defaultProps.setPushNotifications).toHaveBeenCalledWith(device.device_id, !enabled);
    });

    it("changes the local notifications settings status when clicked", () => {
        const device = {
            ...baseDevice,
        };

        const enabled = false;

        const { getByTestId } = render(
            getComponent({
                device,
                localNotificationSettings: {
                    is_silenced: !enabled,
                },
                isSigningOut: true,
            }),
        );

        const checkbox = getByTestId("device-detail-push-notification-checkbox");
        fireEvent.click(checkbox);

        expect(defaultProps.setPushNotifications).toHaveBeenCalledWith(device.device_id, !enabled);
    });
});
