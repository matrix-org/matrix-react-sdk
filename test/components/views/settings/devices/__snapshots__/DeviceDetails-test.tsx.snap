// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeviceDetails /> renders a verified device 1`] = `
<div>
  <div
    class="mx_DeviceDetails"
    data-testid="device-detail-my-device"
  >
    <section
      class="mx_DeviceDetails_section"
    >
      <h3
        class="mx_Heading_h3"
      >
        my-device
      </h3>
      <div
        class="mx_DeviceSecurityCard"
      >
        <div
          class="mx_DeviceSecurityCard_icon Verified"
        >
          <div
            height="16"
            width="16"
          />
        </div>
        <div
          class="mx_DeviceSecurityCard_content"
        >
          <p
            class="mx_DeviceSecurityCard_heading"
          >
            Verified session
          </p>
          <p
            class="mx_DeviceSecurityCard_description"
          >
            This session is ready for secure messaging.
          </p>
        </div>
      </div>
    </section>
    <section
      class="mx_DeviceDetails_section"
    >
      <p
        class="mx_DeviceDetails_sectionHeading"
      >
        Session details
      </p>
      <table
        class="mxDeviceDetails_metadataTable"
      >
        <tbody>
          <tr>
            <td
              class="mxDeviceDetails_metadataLabel"
            >
              Session ID
            </td>
            <td
              class="mxDeviceDetails_metadataValue"
            >
              my-device
            </td>
          </tr>
          <tr>
            <td
              class="mxDeviceDetails_metadataLabel"
            >
              Last activity
            </td>
            <td
              class="mxDeviceDetails_metadataValue"
            />
          </tr>
        </tbody>
      </table>
      <table
        class="mxDeviceDetails_metadataTable"
      >
        <thead>
          <tr>
            <th>
              Device
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="mxDeviceDetails_metadataLabel"
            >
              IP address
            </td>
            <td
              class="mxDeviceDetails_metadataValue"
            />
          </tr>
        </tbody>
      </table>
    </section>
  </div>
</div>
`;

exports[`<DeviceDetails /> renders device with metadata 1`] = `
<div>
  <div
    class="mx_DeviceDetails"
    data-testid="device-detail-my-device"
  >
    <section
      class="mx_DeviceDetails_section"
    >
      <h3
        class="mx_Heading_h3"
      >
        My Device
      </h3>
      <div
        class="mx_DeviceSecurityCard"
      >
        <div
          class="mx_DeviceSecurityCard_icon Unverified"
        >
          <div
            height="16"
            width="16"
          />
        </div>
        <div
          class="mx_DeviceSecurityCard_content"
        >
          <p
            class="mx_DeviceSecurityCard_heading"
          >
            Unverified session
          </p>
          <p
            class="mx_DeviceSecurityCard_description"
          >
            Verify or sign out from this session for best security and reliability.
          </p>
        </div>
      </div>
    </section>
    <section
      class="mx_DeviceDetails_section"
    >
      <p
        class="mx_DeviceDetails_sectionHeading"
      >
        Session details
      </p>
      <table
        class="mxDeviceDetails_metadataTable"
      >
        <tbody>
          <tr>
            <td
              class="mxDeviceDetails_metadataLabel"
            >
              Session ID
            </td>
            <td
              class="mxDeviceDetails_metadataValue"
            >
              my-device
            </td>
          </tr>
          <tr>
            <td
              class="mxDeviceDetails_metadataLabel"
            >
              Last activity
            </td>
            <td
              class="mxDeviceDetails_metadataValue"
            >
              Sun 22:34
            </td>
          </tr>
        </tbody>
      </table>
      <table
        class="mxDeviceDetails_metadataTable"
      >
        <thead>
          <tr>
            <th>
              Device
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="mxDeviceDetails_metadataLabel"
            >
              IP address
            </td>
            <td
              class="mxDeviceDetails_metadataValue"
            >
              123.456.789
            </td>
          </tr>
        </tbody>
      </table>
    </section>
  </div>
</div>
`;

exports[`<DeviceDetails /> renders device without metadata 1`] = `
<div>
  <div
    class="mx_DeviceDetails"
    data-testid="device-detail-my-device"
  >
    <section
      class="mx_DeviceDetails_section"
    >
      <h3
        class="mx_Heading_h3"
      >
        my-device
      </h3>
      <div
        class="mx_DeviceSecurityCard"
      >
        <div
          class="mx_DeviceSecurityCard_icon Unverified"
        >
          <div
            height="16"
            width="16"
          />
        </div>
        <div
          class="mx_DeviceSecurityCard_content"
        >
          <p
            class="mx_DeviceSecurityCard_heading"
          >
            Unverified session
          </p>
          <p
            class="mx_DeviceSecurityCard_description"
          >
            Verify or sign out from this session for best security and reliability.
          </p>
        </div>
      </div>
    </section>
    <section
      class="mx_DeviceDetails_section"
    >
      <p
        class="mx_DeviceDetails_sectionHeading"
      >
        Session details
      </p>
      <table
        class="mxDeviceDetails_metadataTable"
      >
        <tbody>
          <tr>
            <td
              class="mxDeviceDetails_metadataLabel"
            >
              Session ID
            </td>
            <td
              class="mxDeviceDetails_metadataValue"
            >
              my-device
            </td>
          </tr>
          <tr>
            <td
              class="mxDeviceDetails_metadataLabel"
            >
              Last activity
            </td>
            <td
              class="mxDeviceDetails_metadataValue"
            />
          </tr>
        </tbody>
      </table>
      <table
        class="mxDeviceDetails_metadataTable"
      >
        <thead>
          <tr>
            <th>
              Device
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="mxDeviceDetails_metadataLabel"
            >
              IP address
            </td>
            <td
              class="mxDeviceDetails_metadataValue"
            />
          </tr>
        </tbody>
      </table>
    </section>
  </div>
</div>
`;
